rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,296 @@
+---
+authors: Russell Jones (rjones@goteleport.com) and Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools like `tsh` and `tctl` can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools like `tsh` and `tctl` will automatically download and install the
+required version for the Teleport cluster.
+
+Enrollment in automatic updates for client tools will be controlled at the
+cluster level. By default all Cloud clusters will be opted into automatic
+updates for client tools. Cluster administrators using MDM software like Jamf
+will be able opt-out manually manage updates.
+
+Self-hosted clusters will be be opted out, but have the option to use the same
+automatic update mechanism.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
```

Comment:
Could we build some kind of observability around this and use this as a success criteria?

- for the users, so they know which versions their users are using. For example recording the teleport client version on login, and when an update happens.
- for cloud operations, so we know what's the tenant state and can send the apropriate communications/do the right maintenances
- for product decision so we can measure the tool update adoption for customers with dial-home:
  - report the versions of teleport tools
  - report if automatic tool update is enabled
  - report the OS distribution (mac/windows/linux, and which major version)

Reply:
Yes, definitely planning on adding observability. I'll update the RFD a bit later in development on exact metrics, but all of the above sound reasonable.

We'll have to do something to avoid duplicates and preserve anonymity, we can follow a similar approach to what we do for product metrics.